20231214-listUsers-api.md

```diff
@@ -0,0 +1,889 @@
+
+* **Userset** - A userset refers to a set of zero or more concrete objects that are the subjects that a relationship expands to. See our Concepts page on [Users](https://openfga.dev/docs/concepts#what-is-a-user) for more info.
+
+
+## Motivation
+[motivation]: #motivation
+
+Developers using OpenFGA want to look up all of the users that have a particular relationship with an object. Currently, the Read API is inadequate for listing users because it can only display users who have direct access to an object but not with users that have derived access through expansion.
+
+### Use Cases
+
+- **UIs** - Display the users that a resource has been shared with. Ex: the "Share" dialog in Google Docs.
+- **Notifications** - Notify all users who relate to a specific object. Ex: send and email to editors of a document when it has been updated.
+
+### Expected Outcome
+
+Two new core APIs are added to the [OpenFGA API](https://github.com/openfga/api) and implementations of these new APIs in the [OpenFGA server](https://github.com/openfga/openfga).
+
+## What it is
+[what-it-is]: #what-it-is
+
+Given an `object`, `relation`, and one or more user/subject provided filters, return the concrete/terminal users or sets of users matching at least one of the user filters that have that relationship with the object.
+
+More formally, given the following model and relationship tuples:
+
+```
+model
+  schema 1.1
+
+type user
+type cat
+
+type group
+  relations
+    define member: [user, group#member]
+
+type document
+  relations
+    define viewer: [cat, user, group#member]
+```
+
+| object     | relation | user             | condition_name | context |
+|------------|----------|------------------|----------------|---------|
+| document:1 | viewer   | user:anne        | null           | null    |
+| document:1 | viewer   | group:eng#member | null           | null    |
+| group:eng  | member   | group:fga#member | null           | null    |
+| group:fga  | member   | user:jon         | null           | null    |
+
+ListUsers and StreamedListUsers would behave as follows:
+
+> ℹ️ For the sake of brevity the examples below focus on the ListUsers unary RPC variant. The difference with the StreamedListUsers API is the streaming semantics of the response. Instead of returning an array, the streaming variant returns each result one at a time as they are "discovered".
+
+**Example 1**
+`user` object types are both directly and indirectly related to `document#viewer` relationships. The indirect relationships come through the direct relationships involving `group#member` (e.g. users can directly view a document or if the user is part of a group that can view the document)
+```
+ListUsers({
+    object: "document:1",
+    relation: "viewer",
+    user_filters: [{type: "user"}]
+}) --> 
+{
+  users: [],
+  wildcard_types: ["user:anne", "user:jon"]
```

Comment:
Should this response be in the `users` array, since the wildcard_types do not tie to an actual user? We don't return user objects in any other wildcard_types examples.